### ISSUE TITLE
[Bug]: Inpaint mask for text2img API doesn't work

### ISSUE THREAD

**Issue by tiimgreen** [Mikubill/sd-webui-controlnet#2242]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

When you pass the image mask as a base64 encoded string to the controlnet params for the API it has no effect.

Everything runs fine but its as if no mask was passed

This is also mentioned here: #2006

For example, like this, the `encoded_mask` has no effect

```python
payload = {
    ...
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "enabled": True,
                    "input_image": [encoded_image, encoded_mask],
                    "model": "control_v11p_sd15_inpaint [ebff9138]",
                    "module": "inpaint_only+lama",
                    "control_mode": 2,
                    "pixel_perfect": True
                }
            ]
        }
    }
}
```

I have tried white on black and black on white masks

### Steps to reproduce the problem

1. Start API server for webui with controlnet
2. Send the above payload

### What should have happened?

It should have inpainted the masked area

### Commit where the problem happens

webui: 4afaaf8a020c1df457bcf7250cb1c7f609699fa7
controlnet: e382d1618593ae05a7115006de8680d3ddbd9777

### Command Line Arguments

```Shell
python /stable-diffusion-webui/webui.py \
  --skip-python-version-check \
  --skip-torch-cuda-test \
  --skip-install \
  --skip-version-check \
  --ckpt /stable-diffusion-webui/models/Stable-diffusion/model.safetensors \
  --opt-sdp-attention \
  --disable-safe-unpickle \
  --port 3000 \
  --api \
  --nowebui \
  --no-hashing \
  --no-download-sd-model \
  --lowram \
  --no-half
```

### List of enabled extensions

- controlnet

### Console logs

```Shell
Worker Initiated
Starting WebUI API
Starting RunPod Handler
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
no module 'xformers'. Processing without...
no module 'xformers'. Processing without...
Service not ready yet. Retrying...
No module 'xformers'. Proceeding without it.
Warning: caught exception 'Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx', memory monitor disabled
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
Service not ready yet. Retrying...
2023-11-12 13:25:27,997 - ControlNet - [0;32mINFO[0m - ControlNet v1.1.411
ControlNet preprocessor location: /stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/downloads
2023-11-12 13:25:28,097 - ControlNet - [0;32mINFO[0m - ControlNet v1.1.411
Service not ready yet. Retrying...
Loading weights [None] from /stable-diffusion-webui/models/Stable-diffusion/model.safetensors
Service not ready yet. Retrying...
Startup time: 7.0s (import torch: 3.1s, import gradio: 0.7s, setup paths: 1.2s, initialize shared: 0.1s, other imports: 0.8s, setup codeformer: 0.1s, load scripts: 0.7s).
[32mINFO[0m:     Started server process [[36m7[0m]
[32mINFO[0m:     Waiting for application startup.
[32mINFO[0m:     Application startup complete.
[32mINFO[0m:     Uvicorn running on [1mhttp://127.0.0.1:3000[0m (Press CTRL+C to quit)
[32mINFO[0m:     127.0.0.1:46992 - "[1mGET /txt2img HTTP/1.1[0m" [31m404 Not Found[0m
WebUI API Service is ready. Starting RunPod...
--- Starting Serverless Worker |  Version 1.3.4.dev8+g34ac99c ---
INFO   | Using test_input.json as job input.
DEBUG  | Retrieved local job: {'input': {'input_image_url': 'https://storage.googleapis.com/snapshot-run-images/3ba4c54d-82d0-4aa4-ac40-da79ee2c5db6/5c94e12a-0a92-4292-9822-1bac20a6e4b2.png', 'mask_image_url': 'https://i.ibb.co/Dk8kHsF/zimmer-mask.png'}, 'id': 'local_test'}
INFO   | local_test | Started
Fetching image: https://storage.googleapis.com/snapshot-run-images/3ba4c54d-82d0-4aa4-ac40-da79ee2c5db6/5c94e12a-0a92-4292-9822-1bac20a6e4b2.png
Creating model from config: /stable-diffusion-webui/configs/v1-inference.yaml

Downloading (…)olve/main/vocab.json:   0% 0.00/961k [00:00<?, ?B/s]
Downloading (…)olve/main/vocab.json: 100% 961k/961k [00:00<00:00, 3.26MB/s]
Downloading (…)olve/main/vocab.json: 100% 961k/961k [00:00<00:00, 3.25MB/s]
Fetching image: https://i.ibb.co/Dk8kHsF/zimmer-mask.png
Downloading (…)olve/main/merges.txt:   0% 0.00/525k [00:00<?, ?B/s]
Downloading (…)olve/main/merges.txt: 100% 525k/525k [00:00<00:00, 8.47MB/s]
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git

Downloading (…)cial_tokens_map.json:   0% 0.00/389 [00:00<?, ?B/s]
Downloading (…)cial_tokens_map.json: 100% 389/389 [00:00<00:00, 2.77MB/s]

Downloading (…)okenizer_config.json:   0% 0.00/905 [00:00<?, ?B/s]
Downloading (…)okenizer_config.json: 100% 905/905 [00:00<00:00, 6.32MB/s]

Downloading (…)lve/main/config.json:   0% 0.00/4.52k [00:00<?, ?B/s]
Downloading (…)lve/main/config.json: 100% 4.52k/4.52k [00:00<00:00, 27.0MB/s]
Applying attention optimization: sdp... done.
Model loaded in 4.1s (load weights from disk: 0.3s, create model: 2.2s, apply weights to model: 0.8s, apply float(): 0.4s, calculate empty prompt: 0.3s).
2023-11-12 13:25:32,641 - ControlNet - [0;32mINFO[0m - Loading model: control_v11p_sd15_inpaint [ebff9138]
2023-11-12 13:25:33,588 - ControlNet - [0;32mINFO[0m - Loaded state_dict from [/stable-diffusion-webui/extensions/sd-webui-controlnet/models/control_v11p_sd15_inpaint.pth]
2023-11-12 13:25:33,589 - ControlNet - [0;32mINFO[0m - controlnet_default_config
2023-11-12 13:25:35,893 - ControlNet - [0;32mINFO[0m - ControlNet model control_v11p_sd15_inpaint [ebff9138] loaded.
2023-11-12 13:25:35,911 - ControlNet - [0;32mINFO[0m - using inpaint as input
2023-11-12 13:25:35,921 - ControlNet - [0;32mINFO[0m - Loading preprocessor: inpaint_only+lama
2023-11-12 13:25:35,921 - ControlNet - [0;32mINFO[0m - preprocessor resolution = 512
Downloading: "https://huggingface.co/lllyasviel/Annotators/resolve/main/ControlNetLama.pth" to /stable-diffusion-webui/extensions/sd-webui-controlnet/annotator/downloads/lama/ControlNetLama.pth


  0% 0.00/195M [00:00<?, ?B/s]
 13% 24.4M/195M [00:00<00:00, 256MB/s]
 28% 55.4M/195M [00:00<00:00, 296MB/s]
 44% 85.6M/195M [00:00<00:00, 306MB/s]
 60% 116M/195M [00:00<00:00, 312MB/s] 
 75% 147M/195M [00:00<00:00, 316MB/s]
 92% 179M/195M [00:00<00:00, 323MB/s]
100% 195M/195M [00:00<00:00, 315MB/s]
2023-11-12 13:25:42,659 - ControlNet - [0;32mINFO[0m - ControlNet used torch.float32 VAE to encode torch.Size([1, 4, 64, 64]).
2023-11-12 13:25:42,684 - ControlNet - [0;32mINFO[0m - ControlNet Hooked - Time = 10.26375937461853
2023-11-12 13:25:42,894 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

  0% 0/20 [00:00<?, ?it/s]

Total progress:   0% 0/20 [00:00<?, ?it/s][A2023-11-12 13:25:51,136 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

  5% 1/20 [00:08<02:36,  8.24s/it]

Total progress:  10% 2/20 [00:08<01:12,  4.05s/it][A2023-11-12 13:25:59,234 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 10% 2/20 [00:16<02:26,  8.16s/it]

Total progress:  15% 3/20 [00:16<01:37,  5.76s/it][A2023-11-12 13:26:07,383 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 15% 3/20 [00:24<02:18,  8.15s/it]

Total progress:  20% 4/20 [00:24<01:46,  6.63s/it][A2023-11-12 13:26:15,474 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 20% 4/20 [00:32<02:10,  8.13s/it]

Total progress:  25% 5/20 [00:32<01:47,  7.17s/it][A2023-11-12 13:26:23,650 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 25% 5/20 [00:40<02:02,  8.15s/it]

Total progress:  30% 6/20 [00:40<01:44,  7.46s/it][A2023-11-12 13:26:31,704 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 30% 6/20 [00:48<01:53,  8.11s/it]

Total progress:  35% 7/20 [00:48<01:40,  7.70s/it][A2023-11-12 13:26:39,908 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 35% 7/20 [00:57<01:45,  8.14s/it]

Total progress:  40% 8/20 [00:56<01:33,  7.81s/it][A2023-11-12 13:26:47,952 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 40% 8/20 [01:05<01:37,  8.11s/it]

Total progress:  45% 9/20 [01:04<01:26,  7.89s/it][A2023-11-12 13:26:56,015 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 45% 9/20 [01:13<01:29,  8.10s/it]

Total progress:  50% 10/20 [01:12<01:19,  7.93s/it][A2023-11-12 13:27:04,037 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 50% 10/20 [01:21<01:20,  8.07s/it]

Total progress:  55% 11/20 [01:20<01:11,  7.97s/it][A2023-11-12 13:27:12,115 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 55% 11/20 [01:29<01:12,  8.08s/it]

Total progress:  60% 12/20 [01:29<01:03,  7.99s/it][A2023-11-12 13:27:20,149 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 60% 12/20 [01:37<01:04,  8.06s/it]

Total progress:  65% 13/20 [01:37<00:56,  8.03s/it][A2023-11-12 13:27:28,263 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 65% 13/20 [01:45<00:56,  8.08s/it]

Total progress:  70% 14/20 [01:45<00:48,  8.04s/it][A2023-11-12 13:27:36,325 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 70% 14/20 [01:53<00:48,  8.07s/it]

Total progress:  75% 15/20 [01:53<00:40,  8.10s/it][A2023-11-12 13:27:44,564 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 75% 15/20 [02:01<00:40,  8.12s/it]

Total progress:  80% 16/20 [02:01<00:32,  8.10s/it][A2023-11-12 13:27:52,658 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 80% 16/20 [02:09<00:32,  8.11s/it]

Total progress:  85% 17/20 [02:09<00:24,  8.08s/it][A2023-11-12 13:28:00,704 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 85% 17/20 [02:17<00:24,  8.09s/it]

Total progress:  90% 18/20 [02:17<00:16,  8.10s/it][A2023-11-12 13:28:08,830 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 90% 18/20 [02:25<00:16,  8.10s/it]

Total progress:  95% 19/20 [02:25<00:08,  8.10s/it][A2023-11-12 13:28:16,933 - ControlNet - [0;32mINFO[0m - [ControlNet] Initial noise hack applied to torch.Size([1, 4, 64, 64]).

 95% 19/20 [02:34<00:08,  8.10s/it]

Total progress: 100% 20/20 [02:33<00:00,  8.11s/it][A
100% 20/20 [02:42<00:00,  8.12s/it]
100% 20/20 [02:42<00:00,  8.11s/it]


Total progress: 100% 20/20 [02:41<00:00,  8.11s/it][A
Total progress: 100% 20/20 [02:41<00:00,  8.08s/it]
[32mINFO[0m:     127.0.0.1:47008 - "[1mPOST /sdapi/v1/txt2img HTTP/1.1[0m" [32m200 OK[0m

// prints base 64 output which is just the original image

INFO   | Local testing complete, exiting.
```

### Additional information

This is a simplified version of a python file I have which looks something like this:
```
def get_as_base64(url):
    print("Fetching image: " + url)
    return base64.b64encode(requests.get(url).content).decode('utf-8')

def run_inference():
    encoded_image = get_as_base64("https://storage.googleapis.com/snapshot-run-images/3ba4c54d-82d0-4aa4-ac40-da79ee2c5db6/5c94e12a-0a92-4292-9822-1bac20a6e4b2.png")
    encoded_mask = get_as_base64("https://i.ibb.co/Dk8kHsF/zimmer-mask.png")

    payload = {
        "prompt": "",
        "negative_prompt": "",
        "sampler_name": "Euler a",
        "batch_size": 1,
        "steps": 20,
        "cfg_scale": 7,
        "width": 512,
        "height": 512,
        "alwayson_scripts": {
            "controlnet": {
                "args": [
                    {
                        "enabled": True,
                        "input_image": [encoded_image, encoded_mask],
                        "model": "control_v11p_sd15_inpaint [ebff9138]",
                        "module": "inpaint_only+lama",
                        "control_mode": 2, # controlnet mor important
                        "pixel_perfect": True
                    }
                ]
            }
        }
    }

    response = automatic_session.post(
        url=f'{SD_API_URL}/txt2img',
        json=payload,
        timeout=600
    )

    images = response.json()["images"]

    return images[0]

run_inference()
```

So I start the SD webui with the above options and then run the python file and it just gives back the same input image, as if the mask was ignored.

**Comment by tiimgreen** [comment]
@huchenlei this is a big blocker for me, would love to hear your thoughts on this

**Comment by huchenlei** [comment]
Can you share the images you used?
- ControlNet input image
- Mask image
- Result image you get

Also can you share the full console logs?

**Comment by tiimgreen** [comment]
@huchenlei sure, just added more details with steps to reproduce and logs. Here is the output in base 64 (which is the same as the input image) as if the mask was ignored: https://anotepad.com/notes/qqfedgyw

**Comment by tiimgreen** [comment]
@huchenlei does the above help?

**Comment by huchenlei** [comment]
You can try replacing
```js
"input_image": [encoded_image, encoded_mask],
```
with
```js
"image": {"image": encoded_image, "mask": encoded_mask},
```

Multiple input formats are accepted by ControlNet API. I am using the dict format in my own project and it is working.

**Comment by IgorAherne** [comment]
where you were doing:
```
"alwayson_scripts": {
            "controlnet": {
                "args": [
                    {
                        "enabled": True,
                        "input_image": base64imageString,
                        "mask":  base64mask_blackAndWhite,  #no alpha,, either perfect white or perfect black
                        "model": "control_v11p_sd15_inpaint [ebff9138]",
                        "module": "inpaint_only+lama",
                        "control_mode": 2, # controlnet more important
                        "pixel_perfect": True
                    }
                ]
            }
        }
```
Here are more about args https://github.com/Mikubill/sd-webui-controlnet/wiki/API#controlnetunitrequest-json-object

**Comment by AIGC404** [comment]
@huchenlei 
I have encountered the same issue. When using the API to generate results with a fixed seed, the output differs from what is obtained through the web UI. The API's mask parameter was obtained by capturing network traffic using the browser's F12 developer tools, so I am certain that all the parameters are guaranteed to be 100% identical.

I have noticed that the results of the Controlnet preprocessor obtained through the API are inconsistent with those obtained through the web UI. This issue is likely the cause of the discrepancy.

Here are my parameters:
 {
    "pixel_perfect": false,
    "control_mode": 0,
    "module": "inpaint_only",
    "image": {
        "image": "...Image base64",
        "mask": "...Mask base64"
    },
    "weight": 1,
    "model": "control_v11p_sd15_inpaint [ebff9138]",
    "enabled": false
}
